pkg/sitewise/datasource.go

```diff
@@ -23,13 +25,56 @@ type Datasource struct {
 }
 
 func NewDatasource(settings backend.DataSourceInstanceSettings) (*Datasource, error) {
-	cfg := client.AWSSiteWiseDataSourceSetting{}
+	cfg := models.AWSSiteWiseDataSourceSetting{}
+
 	err := cfg.Load(settings)
 	if err != nil {
 		return nil, err
 	}
-	sessions := awsds.NewSessionCache()
 
+	err = cfg.Validate()
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Region == models.EDGE_REGION && cfg.EdgeAuthMode != models.EDGE_AUTH_MODE_DEFAULT {
+		// TODO: refresh session every 4h since creds expire
+		edgeAuthenticator := EdgeAuthenticator{
+			Settings: cfg,
+		}
+
+		var waitTime time.Duration
+
+		updateAuth := func() error {
+			authInfo, err := edgeAuthenticator.Authorize()
+			if err == nil {
+				cfg.AccessKey = authInfo.AccessKeyId
+				cfg.SecretKey = authInfo.SecretAccessKey
+				cfg.SessionToken = authInfo.SessionToken
+				cfg.AuthType = awsds.AuthTypeKeys
+				waitTime = time.Until(authInfo.SessionExpiryTime)
+				log.DefaultLogger.Debug("should wait for: ", "time:", waitTime)
+				//waitTime = 10 * time.Second
+			}
+			return err
+		}
+
+		err = updateAuth()
+		if err != nil {
+			return &Datasource{}, err
+		}
+
+		go func() {
+			for {
+				log.DefaultLogger.Debug("wait time until next credential fetch: ", "time:", waitTime)
+				<-time.After(waitTime)
+				log.DefaultLogger.Debug("updating edge auth credentials now")
+				updateAuth()
```

Comment:
Maybe in case of error here we should try to update with some shorter interval?

Reply:
btw, one more option to consider - avoid goroutine at all, do `updateAuth()` on initialization, save `authInfo.SessionExpiryTime`, then inside `GetClient` check  SessionExpiryTime and try to `updateAuth` again if needed.

Reply:
yeah i think for now an else with a small waitTime would be okay - although the other approach seems simpler